piptools/repositories/local.py

```diff
@@ -76,10 +77,9 @@ def find_best_match(
         key = key_from_ireq(ireq)
         existing_pin = self.existing_pins.get(key)
         if existing_pin and ireq_satisfied_by_existing_pin(ireq, existing_pin):
-            project, version, _ = as_tuple(existing_pin)
-            return make_install_requirement(project, version, ireq)
-        else:
-            return self.repository.find_best_match(ireq, prereleases)
+            with suppress(NoCandidateFound):
+                return self.repository.find_best_match(existing_pin, prereleases)
```

Comment:
BTW we can also suppress yanked releases if `best_candidate.link.is_yanked is True`. What do you think?

Reply:
Also I think we should print a warning here that existing pin is not valid/exists.

Reply:
Yes, thanks, we shouldn't use yanked releases -- I didn't realize `find_best_match` would return them. I tested with a found yanked release (input: `gssapi`; output: `gssapi==1.6.6`), and unfortunately in this case the `link` attribute of the `InstallRequirement` candidate, as returned by `find_best_match`, is `None`, so we can't immediately check `candidate.link.is_yanked`.

It's been just over a year since I submitted this PR, and I don't readily recall how we mutate `InstallRequirements` to fill the attributes with useful info, if that's what's needed.

I also can't find any existing explicit mention/handling/testing of yanked releases in the codebase.

And yes, once we figure out the yanked business, I'll add a logged warning about discarded matches, thanks!

Reply:
I'm now doubting that we should reject already-pinned yanked versions. Despite a package being yanked (not deleted), a downstream package may still need that version, and understand the risks, or want to take time to handle it mindfully.

Another yanked example is `attrs` `21.1.0`, and when that is already pinned during a non-upgrading compile, we get a nice informative warning:

```
WARNING: The candidate selected for download or install is a yanked version: 'attrs' candidate (version 21.1.0 at https://files.pythonhosted.org/packages/2b/00/7f74fb9608daf3626aad5de91181026a5107b997ce2788eabffeb5d41cb1/attrs-21.1.0-py2.py3-none-any.whl (from https://pypi.org/simple/attrs/) (requires-python:>=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*))
  Reason for being yanked: Installable but not importable on Python 3.4.
```

Reply:
As an aside, do you know why running `tox -e checkqa` locally always has isort move `from build import BuildBackendException` in `piptools/scripts/compile.py`, but in CI it moves back?

Reply:
Exactly the problem, thanks!

Reply:
I copied the warning that you're already using in the backtracking resolver.

Reply:
About yanked releases, I'm now thinking:

- existing pins which are yanked should indeed be discarded
- yanked releases must still be valid candidates, allowing for users to constrain their way into using a yanked release if they *need* to
- despite being "valid candidates," they must not prevail over a non-yanked alternative release satisfying the constraints